validator/db/proposal_history.go

```diff
@@ -0,0 +1,101 @@
+package db
+
+import (
+	"github.com/boltdb/bolt"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	slashpb "github.com/prysmaticlabs/prysm/proto/slashing"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// HasProposedForEpoch returns whether a validators proposal history has been marked for the entered epoch.
+// If the request is more in the future than what the history contains, it will return false.
+// If the request is from the past, and likely previously pruned it will return false.
+func HasProposedForEpoch(history *slashpb.ProposalHistory, epoch uint64) bool {
```

Comment:
I think having these functions here is fine, or you could also place them in the validator_propose.go file